qutebrowser/browser/history.py

```diff
@@ -288,14 +202,110 @@ def add_url(self, url, title="", *, redirect=False, atime=None):
+        else:
+            flags = ''
+
+        if not set(flags).issubset('r'):
+            raise ValueError("Invalid flags {!r}".format(flags))
+
+        redirect = 'r' in flags
+        row = (url, title, float(atime), redirect)
+        completion_row = None if redirect else (url, title, float(atime))
+
+        return (row, completion_row)
+
+    def import_txt(self):
+        """Import a history text file into sqlite if it exists.
+
+        In older versions of qutebrowser, history was stored in a text format.
+        This converts that file into the new sqlite format and moves it to a
+        backup location.
+        """
+        path = os.path.join(standarddir.data(), 'history')
+        if not os.path.isfile(path):
+            return
+
+        def action():
+            with debug.log_time(log.init, 'Import old history file to sqlite'):
+                try:
+                    self._read(path)
+                except ValueError as ex:
+                    message.error('Failed to import history: {}'.format(ex))
+                else:
+                    bakpath = path + '.bak'
+                    message.info('History import complete. Moving {} to {}'
+                                 .format(path, bakpath))
+                    os.rename(path, bakpath)
+
+        # delay to give message time to appear before locking down for import
+        message.info('Converting {} to sqlite...'.format(path))
+        QTimer.singleShot(100, action)
+
+    def _read(self, path):
+        """Import a text file into the sql database."""
+        with open(path, 'r', encoding='utf-8') as f:
+            rows = []
+            completion_rows = []
+            for (i, line) in enumerate(f):
+                line = line.strip()
+                if not line:
+                    continue
+                try:
+                    row, completion_row = self._parse_entry(line.strip())
+                    rows.append(row)
+                    if completion_row is not None:
+                        completion_rows.append(completion_row)
+                except ValueError as ex:
+                    raise ValueError('Failed to parse line #{} of {}: "{}"'
+                                     .format(i, path, ex))
+        self.insert_batch(rows)
+        self.completion.insert_batch(completion_rows, replace=True)
+
+    @cmdutils.register(instance='web-history', debug=True)
+    def debug_dump_history(self, dest):
+        """Dump the history to a file in the old pre-SQL format.
+
+        Args:
+            dest: Where to write the file to.
+        """
+        dest = os.path.expanduser(dest)
+
+        dirname = os.path.dirname(dest)
+        if not os.path.exists(dirname):
+            raise cmdexc.CommandError('Path does not exist', dirname)
+
+        lines = ('{}{} {} {}'
+                 .format(int(x.atime), '-r' * x.redirect, x.url, x.title)
+                 for x in self.select(sort_by='atime', sort_order='asc'))
```

Comment:
Wait... Why do we suddenly get a `MemoryError` [on Travis](https://travis-ci.org/qutebrowser/qutebrowser/jobs/241039298#L1135) here?! 😕